kvdb/cache.go

```diff
@@ -0,0 +1,930 @@
+package kvdb
+
+import (
+	"container/list"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+	"github.com/google/btree"
+)
+
+const (
+	treeDeg = 3
+)
+
+type cacheBucket struct {
+	seq  *uint64
+	tree *btree.BTree
+}
+
+func newCacheBucket() *cacheBucket {
+	return &cacheBucket{
+		tree: btree.New(treeDeg),
+	}
+}
+
+func (c *cacheBucket) get(key []byte) *cachedItem {
+	if key == nil {
+		return nil
+	}
+
+	keyItem := &cachedItem{
+		key: string(key),
+	}
+
+	if valItem := c.tree.Get(keyItem); valItem != nil {
+		return valItem.(*cachedItem)
+	}
+
+	return nil
+}
+
+func (c *cacheBucket) put(item *cachedItem) {
+	c.tree.ReplaceOrInsert(item)
+}
+
+func (c *cacheBucket) del(item *cachedItem) {
+	c.tree.Delete(item)
+}
+
+type cachedItem struct {
+	key    string
+	value  string
+	bucket *cacheBucket
+}
+
+func (c *cachedItem) Less(than btree.Item) bool {
+	return c.key < than.(*cachedItem).key
+}
+
+// Enforce that Cache implements the ExtendedBackend interface.
+var _ walletdb.DB = (*Cache)(nil)
+
+type pendingChange interface {
+	// Reverts the pending change.
+	Revert()
+}
+
+type pendingAdd struct {
+	parent   *cachedItem
+	newChild *cachedItem
+}
+
+func (p *pendingAdd) Revert() {
+	p.parent.bucket.tree.Delete(p.newChild)
```

Comment:
Perhaps add a method on the `cachedItem` itself to handle this so the struct doesn't need to dig so far into the implementation details?

Reply:
The `cachedItem` currently doesn't know about its parent, but we could change that if need to be.